### PR TITLE
Improve perf of j/l/Class annotation APIs

### PR DESCRIPTION
**Pull request by andrewcraik**
The current implementation of the various annoation APIs in
java/lang/Class are based on the AnnotationCache data structure which
has two fields of type Map - an interface type. The maps are either set
to and instance of the EmptyMap type or an instance of LinkedHashMap.
This change switches EmptyMap cases to be null. This in turn allows the
fields to be declared to be of the concrete LinkedHashMap type which
boosts performance and simplifies handling of empty maps. This also
makes the call profiling from the IProfiler much less confusing for the
JIT leading to throughput boost of around 30% in microbenchmarks of the
getAnnotation API.

### CODE REVIEW

**Comment by DanHeidinga** [comment]
@pshipton Can you review this?

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 1]
```diff
@@ -2390,8 +2396,11 @@ static ClassLoader currentClassLoader() {
  * @since 1.5
  */
 public Annotation[] getAnnotations() {
-	Collection<Annotation> annotations = getAnnotationCache().annotationMap.values();
-	return annotations.toArray(new Annotation[annotations.size()]);
+	LinkedHashMap<Class<? extends Annotation>, Annotation> map = getAnnotationCache().annotationMap;
+	if (map != null) {
+		return map.values().toArray(EMPTY_ANNOTATION_ARRAY);
```
Comment (pshipton) [inline jcl/src/java.base/share/classes/java/lang/Class.java]: Passing in the EMPTY_ANNOTATION_ARRAY may be less efficient than passing in an Annotation[] of the correct size. It means the code needs to call Array.newInstance() to create an array of the correct type and size.

**Comment by andrewcraik** [comment]
@pshipton you are right - I actually did this change in two steps and in the first I kept the empty collection and so passing the empty array was more efficient - if there were no annotations you saved allocating a new array by passing the canonical empty one which boosted perf. Now that the code makes the empty case clear (eg null) we can go back to the original implementation to boost perf further. I have updated the change set.

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 2]
```diff
@@ -2460,8 +2474,12 @@ public String getTypeName() {
  * @since 1.5
  */
 public Annotation[] getDeclaredAnnotations() {
-	Collection<Annotation> directAnnotations = getAnnotationCache().directAnnotationMap.values();
-	return directAnnotations.toArray(new Annotation[directAnnotations.size()]);
+	LinkedHashMap<Class<? extends Annotation>, Annotation> map = getAnnotationCache().directAnnotationMap;
+	if (map != null) {
+		Collection<Annotation> annotations = map.values();
+		return annotations.toArray(new Annotation[annotations.size()]);
```
Comment (pshipton) [inline jcl/src/java.base/share/classes/java/lang/Class.java]: original code was directAnnotations.size()

**Pull request merged**